### PR TITLE
Issue #765 gwt gwt exchanges

### LINKED ISSUE #765: Add GWTGWT package
Steps:

* Create abstract base class "Exchange" from GWFGWF package. 
* Might be able to do that for GWFGWT as well if very similar. (Not sure anymore if coupling is cellid based or based on modelnames, if the latter, the data is too different to copy package class 1on1).
* Implement GWTGWT package.

Note: GWTGWT uses the same geometric constants as the GWFGWF package.

### PR DESCRIPTION
**Pull request by luitjansl**
Fixes #765 

# Description
creates gwtgwt exchange packages when splitting a simulation that contains 1 or more transport models. A gwtgwt package is created per partition boundary and per transport model. (while the geometry constants are the same for all transport models, the options in the gwtgwt file are specific to single transport models. However, for now, these settings are not set; this will be the topic of #747
Also adds tests.

 Checklist

- [X] Links to correct issue
- [ ] Update changelog, if changes affect users
- [X] PR title starts with ``Issue #nr``, e.g. ``Issue #737``
- [X] Unit tests were added
- [ ] **If feature added**: Added/extended example

### CODE REVIEW

File: imod/mf6/gwtgwt.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+from imod.mf6.package import Package
+from imod.typing import GridDataArray
+
+
+class GWTGWT(ExchangeBase):
+    """
+    This package is for writing an exchange file, used for splitting up a model
+    into different submodels (that can be solved in parallel). It (usually)
+    is not instantiated by users, but created by the "split" method of the
+    simulation class."""
```
Comment (JoerivanEngelen) [inline imod/mf6/gwtgwt.py:18]: Please add a newline here:

```suggestion
    simulation class.
    """
```

File: imod/tests/fixtures/flow_transport_simulation_fixture.py [hunk 2]
```diff
@@ -196,16 +196,16 @@ def flow_transport_simulation():
...
-        inner_rclose=0.001,
-        inner_maximum=100,
+        inner_dvclose=1.0e-6,
+        inner_rclose=0.0001,
+        inner_maximum=200,
         linear_acceleration="bicgstab",
         scaling_method=None,
         reordering_method=None,
-        relaxation_factor=0.97,
+        relaxation_factor=0.9,
```
Comment (Manangka) [inline imod/tests/fixtures/flow_transport_simulation_fixture.py:208]: Why are these changes needed?

File: imod/mf6/gwtgwt.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        self.dataset["cell_id2"] = cell_id2
+        self.dataset["layer"] = layer
+        self.dataset["model_name_1"] = transport_model_id1
+        self.dataset["model_name_2"] = transport_model_id2
+        self.dataset["flow_model_name_1"] = flow_model_id1
+        self.dataset["flow_model_name_2"] = flow_model_id2
+        self.dataset["ihc"] = xr.DataArray(np.ones_like(cl1, dtype=int))
+        self.dataset["cl1"] = cl1
+        self.dataset["cl2"] = cl2
+        self.dataset["hwva"] = hwva
```
Comment (JoerivanEngelen) [inline imod/mf6/gwtgwt.py:50]: You are still using the old way of initializing datasets. For example, for the ``GWFGWF`` this changed to:

```python
        dict_dataset = {
            "cell_id1": cell_id1,
            "cell_id2": cell_id2,
            "layer": layer,
            "model_name_1": model_id1,
            "model_name_2": model_id2,
            "ihc": xr.DataArray(np.ones_like(cl1, dtype=int)),
            "cl1": cl1,
            "cl2": cl2,
            "hwva": hwva,
        }
        super().__init__(dict_dataset)
```

File: imod/mf6/simulation.py [hunk 4]
```diff
@@ -941,12 +941,22 @@ def split(self, submodel_labels: xr.DataArray) -> Modflow6Simulation:
                 new_simulation[solution_name].add_model_to_solution(new_model_name)
 
         exchanges = []
-        for model_name, model in original_models.items():
-            if isinstance(model, GroundwaterFlowModel):
-                exchanges += exchange_creator.create_gwfgwf_exchanges(
-                    model_name, model.domain.layer
+        flow_models = self.get_models_of_type("gwf6")
```
Comment (Manangka) [inline imod/mf6/simulation.py:944]: This is a check that you want to do as early as possible. You can move it to somewhere at the top of this method.

File: imod/mf6/simulation.py [hunk 5]
```diff
@@ -1004,6 +1014,9 @@ def _add_modelsplit_exchanges(self, exchanges_list: list[GWFGWF]) -> None:
     def _set_exchange_options(self):
         # collect some options that we will auto-set
         for exchange in self["split_exchanges"]:
+            if isinstance(exchange, GWFGWT):
+                # that package does not have options
+                continue
```
Comment (JoerivanEngelen) [inline imod/mf6/simulation.py:1019]: This check is not necessary. Please remove.

File: imod/mf6/simulation.py [hunk 6]
```diff
@@ -1014,6 +1027,9 @@ def _set_exchange_options(self):
                     xt3d=model_1["npf"].get_xt3d_option(),
                     newton=model_1.is_use_newton(),
                 )
+            if isinstance(exchange, GWTGWT):
```
Comment (JoerivanEngelen) [inline imod/mf6/simulation.py:1030]: This can be changed to:

```suggestion
            elif isinstance(exchange, GWTGWT):
```

**Review by JoerivanEngelen — CHANGES_REQUESTED** [review]
Looks good, I have a few suggestions. 
Most importantly: The dataset initialization should be changed.

File: imod/mf6/simulation.py [hunk 7]
```diff
@@ -941,12 +941,22 @@ def split(self, submodel_labels: xr.DataArray) -> Modflow6Simulation:
...
-            if isinstance(model, GroundwaterFlowModel):
-                exchanges += exchange_creator.create_gwfgwf_exchanges(
-                    model_name, model.domain.layer
+        flow_models = self.get_models_of_type("gwf6")
+        if len(flow_models) != 1:
+            raise ValueError(
+                "splitting of simulations with more (or less) than 1 flow model currently not supported"
+            )
+        flow_model_name = list(flow_models.keys())[0]
+        flow_model = self[flow_model_name]
```
Comment (Manangka) [inline imod/mf6/simulation.py:950]: Dont know if this is better but you can combine these 2 lines into:
`flow_model_name, flow_model= next(iter(gwf_models.items()))`

File: imod/mf6/simulation.py [hunk 8]
```diff
@@ -941,12 +941,22 @@ def split(self, submodel_labels: xr.DataArray) -> Modflow6Simulation:
...
+        exchanges += exchange_creator.create_gwfgwf_exchanges(
+            flow_model_name, flow_model.domain.layer
+        )
+        transport_models = self.get_models_of_type("gwt6")
+        if len(transport_models) > 0:
```
Comment (Manangka) [inline imod/mf6/simulation.py:955]: Intead of a len check you can use `any`. I think that better expresses your intent
`if any(transport_model):`

**Review by Manangka — APPROVED** [review]

File: imod/mf6/gwtgwt.py (hunk 3 above)
Comment (luitjansl) [reply imod/mf6/gwtgwt.py:51]: done

File: imod/tests/fixtures/flow_transport_simulation_fixture.py (hunk 2 above)
Comment (luitjansl) [reply imod/tests/fixtures/flow_transport_simulation_fixture.py:208]: to achieve a better match between the partitioned model and the unpartitioned model. These force the number of iterations between subdomains to increase, lessening the impact of partitioning on the overall outcome. 
We still get a fairly high inconsistency between the two, and hopefully it will improve once we correctly add settings to the gwtgwt exchange packages.

File: imod/mf6/simulation.py (hunk 8 above)
Comment (luitjansl) [reply imod/mf6/simulation.py:955]: done

File: imod/mf6/simulation.py (hunk 4 above)
Comment (luitjansl) [reply imod/mf6/simulation.py:944]: done

File: imod/mf6/simulation.py (hunk 7 above)
Comment (luitjansl) [reply imod/mf6/simulation.py]: that line is gone.  A test failed in which mulltiple flow models were used in the same simulation, and hence we loop now over the flow models, rather thatn focussing on the first one.

**Review by JoerivanEngelen — APPROVED** [review]
Thanks for improving!
